main.c

```diff
@@ -1854,10 +1915,25 @@ copyup (struct ovl_data *lo, struct ovl_node *node)
 
   if ((st.st_mode & S_IFMT) == S_IFLNK)
     {
-      char p[PATH_MAX + 1];
-      ret = readlinkat (node_dirfd (node), node->path, p, sizeof (p) - 1);
-      if (ret < 0)
-        goto exit;
+      size_t current_size = PATH_MAX + 1;
+      cleanup_free char *p = malloc (current_size);
+
+      while (1)
+        {
+          char *new;
+
+          ret = readlinkat (node_dirfd (node), node->path, p, current_size - 1);
+          if (ret < 0)
+            goto exit;
+          if (ret < current_size - 1)
+            break;
+
+          current_size = current_size * 2;
+          new = realloc (p, current_size);
+          if (new == NULL)
```

Comment:
Need to free p on failure?

Reply:
same here, `p` is marked with `cleanup_free`, so it is freed anytime it is out of scope (and != NULL).